internal/rbd/nodeserver.go

```diff
@@ -338,23 +338,19 @@ func (ns *NodeServer) NodeStageVolume(
 		return &csi.NodeStageVolumeResponse{}, nil
 	}
 
-	transaction := stageTransaction{}
 	// Stash image details prior to mapping the image (useful during Unstage as it has no
 	// voloptions passed to the RPC as per the CSI spec)
 	err = stashRBDImageMetadata(rv, stagingParentPath)
 	if err != nil {
 		return nil, status.Error(codes.Internal, err.Error())
 	}
-	defer func() {
-		if err != nil {
-			ns.undoStagingTransaction(ctx, req, transaction, rv)
-		}
-	}()
 
 	// perform the actual staging and if this fails, have undoStagingTransaction
 	// cleans up for us
-	transaction, err = ns.stageTransaction(ctx, req, cr, rv, isStaticVol)
+	transaction, err := ns.stageTransaction(ctx, req, cr, rv, isStaticVol)
 	if err != nil {
+		ns.undoStagingTransaction(ctx, req, transaction, rv)
```

Comment:
as defer is removed. add a note to call undo for all the error cases after this call? if a new function call is added so that we will call the undoStagingTrasaction? or another option is to pass the `stageTransaction` pointer to both functions so cleanup will be taken care of by undo operation.